naming/README.md

```diff
@@ -0,0 +1,93 @@
+
+- [Proposal] An additional instruction (or another contract) could do the following: mint an NFT for an already existing name and activate ownership for the holder of the NFT only. This would enable easy trading of names.
+
+
+Program Interface:
+
+```
+pub enum NameStatus {
+    Deleted,
+    Active,
+}
+
+pub type ValueType = u16;
+
+#[derive(BorshDeserialize, BorshSerialize)]
+pub struct NameState {
+    status: u8,
+    values: Vec<(ValueType, String)>,
+    owner_pubkey: Option<Pubkey>,
+    owner_name: Option<String>,
+}
+
+#[repr(C)]
+#[derive(Clone, Debug, PartialEq, BorshSerialize, BorshDeserialize)]
+pub enum NameInstruction {
+    /// Add a record for the given name to the naming service.
+    /// The owner can be specified as a pubkey or (exclusively) as another name (the owner then being
+    /// the owner of that name, this can be usefull for domain name extensions).
+    ///
+    /// Accounts expected by this instruction:
+    ///
+    ///   1. `[writable]` The program account derived from the name
+    ///   2. `[signer]` (Optional) The domain name parent account if the given name is a sub-domain
+    ///   3...N `[]` The list of accounts needed to find an owner which is a pubkey or a null
+    AddName {
+        name: String,
+        values: Vec<(ValueType, String)>,
+        owner_pubkey: Option<Pubkey>,
+        owner_name: Option<String>,
+    },
+    /// Change an entry for the given name to the naming service.
+    /// The existing value for a given Valuetype will be overwritten.
+    /// Additionnal given values will be added.
+    ///
+    /// Accounts expected by this instruction:
+    ///
+    ///   1. `[writable]` The program account derived from the name
+    ///   2. `[signer]` The account of the first owner that is a pubkey 
+    ///   3...N `[]` The list of accounts needed to find an owner which is a pubkey or a null
+    ChangeValue {
+        name: String,
+        values: Vec<(ValueType, String)>,
+    },
+    /// Change the owner of the name.
+    /// If the owner of the given name is itself a name, the program will follow the given list of parents
+    /// until finding a pubkey owner.
+    ///
+    /// Accounts expected by this instruction:
+    ///
+    ///   1. `[writable]` The program account derived from the name
+    ///   2. `[signer]` The current owner of the program account derived from the name
+    ///   3...N `[]` The list of accounts needed to find an owner which is a pubkey or a null
+    ChangeOwner {
+        name: String,
+        owner_address: Pubkey,
+    },
+    /// Delete a name (and receive the allocation rent?)
+    ///
+    /// Accounts expected by this instruction:
+    ///
+    ///   1. `[writable]` The program account derived from the name
+    ///   2. `[signer]` The owner of the program account derived from the name
+    ///   3...N `[]` The list of accounts needed to find an owner which is a pubkey or a null
+    DeleteName { name: String },
```

Comment:
What happens to subdomains?

Reply:
Good question, I guess they would be broken until someone manually deletes them.
Either this or DeleteName can delete multiple accounts and walk down a sub-domain chain but this again would be limited by the transaction size.